src/BTree.mo

```diff
@@ -0,0 +1,110 @@
+/// Imperative sequences as B-Trees.
+
+import A "Array";
+import I "Iter";
+import List "List";
+import Option "Option";
+import Order "Order";
+import P "Prelude";
+import Prim "mo:⛔";
+
+module {
+
+  /// Constants we use to shape the tree.
+  /// See https://en.wikipedia.org/wiki/B-tree#Definition
+  module Constants {
+    let MAX_CHILDREN = 4;
+  };
+
+  public type Compare<K> = {
+    compare : (K, K) -> Order.Order
+  };
+
+  public type Data<K, V> = [(K, V)];
+
+  public type Index<K, V> = {
+    data : Data<K, V>;
+    trees : [Tree<K, V>];
+  };
+
+  public type Tree<K, V> = {
+    #index : Index<K, V>;
+    #data : Data<K, V>;
```

Comment:
Can the Tree variant's values be renamed `#node` and `#leaf`? Index makes it sound more like a hash table type of lookup.

Reply:
Sure, except that I'd prefer to call it "internal" and "leaf", since both are kinds of "nodes" in my mind.

"Index" seemed more natural when I was misunderstanding the structure of a B-Tree, and thinking of it as a B+-Tree (I guess?) where the data is only at the leaves; then the internal nodes only store keys, and serve as an index.  

Now, I've adopted the standard B-Tree definition where key-value pairs can be internal too, and the name "index" is more inappropriate.

BTW, as one reference for "standard", I've been looking at [the Rust implementation in `std`](https://github.com/rust-lang/rust/blob/fdca237d5194bf8a1c9b437ebd2114d1c2ba6195/library/alloc/src/collections/btree/node.rs#L9), and the helpful comments in particular.  Hopefully this implementation will be much (much!) shorter, at least initially.

Reply:
Went with "internal" and "leaf".